models/user_setting.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"fmt"
+	"strings"
+
+	"code.gitea.io/gitea/models/db"
+	"code.gitea.io/gitea/modules/setting"
+
+	"xorm.io/builder"
+)
+
+// UserSetting is a key value store of user settings
+type UserSetting struct {
+	ID     int64  `xorm:"pk autoincr"`
+	UserID int64  `xorm:"index unique(key_userid)"`              // to load all of someone's settings
+	Key    string `xorm:"varchar(255) index unique(key_userid)"` // ensure key is always lowercase
+	Value  string `xorm:"text"`
+}
+
+// BeforeInsert will be invoked by XORM before inserting a record
+func (userSetting *UserSetting) BeforeInsert() {
+	userSetting.Key = strings.ToLower(userSetting.Key)
+}
+
+// BeforeUpdate will be invoked by XORM before updating a record
+func (userSetting *UserSetting) BeforeUpdate() {
+	userSetting.Key = strings.ToLower(userSetting.Key)
+}
+
+// BeforeDelete will be invoked by XORM before updating a record
+func (userSetting *UserSetting) BeforeDelete() {
+	userSetting.Key = strings.ToLower(userSetting.Key)
+}
+
+func init() {
+	db.RegisterModel(new(UserSetting))
+}
+
+// GetUserSetting returns specific settings from user
+func GetUserSetting(uid int64, keys []string) ([]*UserSetting, error) {
```

Comment:
I've swapped it to return a map. I would prefer to keep lowercase logic because some databases will treat different cases as the same value and we could run into some weird edge cases otherwise.

Reply:
About the `lowercase`: if we do not want to support uppercase, we should do a check to the keys and return an error when caller passes non-lowercase keys.

As a framework, I think it shouldn't change caller's data unexpectedly, it will cause surprises and many cause potential bugs. 

As a framework user, if I do `set("A", ...)` then I can not get `map=get(); v = map["A"]`, it is really a big surprise.

Returning an error doesn't cost much than converting a string to lowercase, and it forces callers to do pass correct data to the framework.

Reply:
That's a reasonable solution. I'll  return error if non-lowercase chars are used.